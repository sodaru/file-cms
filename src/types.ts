export type AttributeFilter = string[];

export type Content = {
  type: string;
  slug: string;
  content: string;
} & Record<string, unknown>;

export type GetContent = (
  type: string,
  slug: string,
  attributes?: AttributeFilter
) => Promise<Partial<Content>>;

export type FilterProvider = (data: unknown, arg: unknown) => boolean;

export type GeneralExpression = Record<string, unknown>;
export type EqExpression = { eq: string | number | boolean | null };
export type StartsWithExpression = { startsWith: string };
export type EndsWithExpression = { endsWith: string };
export type ContainsExpression = { contains: string };
export type GtExpression = { gt: number };
export type LtExpression = { lt: number };
export type GteExpression = { gte: number };
export type LteExpression = { lte: number };
export type BetweenExpression = { between: [number, number] };
export type BetweenInclusiveExpression = { betweenInclusive: [number, number] };
export type HasExpression = { has: string | number | boolean | null };

export type StringExpression =
  | string
  | EqExpression
  | StartsWithExpression
  | EndsWithExpression
  | ContainsExpression
  | GeneralExpression;

export type TypeExpression = StringExpression;

export type SlugExpression = StringExpression;

export type ContentExpression = StringExpression;

export type AnyExpression =
  | string
  | number
  | boolean
  | null
  | GeneralExpression
  | EqExpression
  | StartsWithExpression
  | EndsWithExpression
  | ContainsExpression
  | GtExpression
  | LtExpression
  | GteExpression
  | LteExpression
  | BetweenExpression
  | HasExpression;

export type ListContent = (
  filters?: {
    type?: TypeExpression;
    slug?: SlugExpression;
    content?: ContentExpression;
  } & Record<string, AnyExpression>,
  attributes?: AttributeFilter,
  customFilterProviders?: Record<string, FilterProvider>
) => Promise<Partial<Content>[]>;
