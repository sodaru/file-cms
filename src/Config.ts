import { NoRootDirSetError } from "./errors/NoRootDirSetError";
import { UnknownFilterProviderError } from "./errors/UnknownFilterProviderError";
import { defaultProviders } from "./internal/filterProviders";
import { FilterProvider } from "./types";

export class Config {
  private static rootDir: string;
  private static filterProviders: Record<string, FilterProvider> = {
    ...defaultProviders
  };

  static setRootDir(rootDir: string) {
    this.rootDir = rootDir;
  }

  static getRootDir() {
    if (typeof this.rootDir == "string") {
      return this.rootDir;
    }
    throw new NoRootDirSetError();
  }

  static setFilterProvider(name: string, provider: FilterProvider) {
    this.filterProviders[name] = provider;
  }

  static getFilterProvider(name: string) {
    if (this.filterProviders[name] === undefined) {
      throw new UnknownFilterProviderError(name);
    }
    return this.filterProviders[name];
  }
}
