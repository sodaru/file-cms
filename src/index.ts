export * from "./types";
export * from "./Config";
export * from "./getContent";
export * from "./listContent";
export * from "./getFile";
