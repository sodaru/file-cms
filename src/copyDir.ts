import { dirname, join, relative } from "path";
import { Config } from "./Config";
import { copyFile, mkdir, readdir, stat } from "fs/promises";

export const copyDir = async (sourceDir: string, targetDir: string) => {
  const rootDir = Config.getRootDir();
  const fullSourceDir = join(rootDir, sourceDir);
  const dirsToCopy = [fullSourceDir];
  while (dirsToCopy.length > 0) {
    const dirToCopy = dirsToCopy.shift();
    const files = await readdir(dirToCopy);
    await Promise.all(
      files.map(async file => {
        const fileOrDirPath = join(dirToCopy, file);
        const stats = await stat(fileOrDirPath);
        if (stats.isDirectory()) {
          dirsToCopy.push(fileOrDirPath);
        } else {
          const targetPath = join(
            targetDir,
            relative(fullSourceDir, fileOrDirPath)
          );
          const targetDirname = dirname(targetPath);
          await mkdir(targetDirname, { recursive: true });
          await copyFile(fileOrDirPath, targetPath);
        }
      })
    );
  }
};
