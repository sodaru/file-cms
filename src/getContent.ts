import { Config } from "./Config";
import { attributeFilter } from "./internal/attributeFilter";
import { fetchContentCached } from "./internal/fetchContentCached";
import { GetContent } from "./types";

export const getContent: GetContent = async (type, slug, attributes) => {
  const rootDir = Config.getRootDir();
  const content = await fetchContentCached(rootDir, type, slug);
  return attributeFilter(content, attributes);
};
