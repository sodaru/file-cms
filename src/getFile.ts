import { join } from "path";
import { Config } from "./Config";
import { readFile } from "fs/promises";
import { FileNotFoundError } from "./errors/FileNotFoundError";

export const getFile = (
  path: string,
  options?: Parameters<typeof readFile>["1"]
) => {
  const rootDir = Config.getRootDir();
  const filePath = join(rootDir, path);
  try {
    return readFile(filePath, options);
  } catch (e) {
    throw new FileNotFoundError(rootDir, path, e);
  }
};
