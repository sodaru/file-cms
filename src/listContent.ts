import path from "path";
import { Config } from "./Config";
import { listDirCached } from "./internal/listDirCached";
import { AnyExpression, Content, ListContent } from "./types";
import { fetchContentCached } from "./internal/fetchContentCached";
import { attributeFilter } from "./internal/attributeFilter";

export const listContent: ListContent = async (
  filters = {},
  attributes = []
) => {
  const rootDir = Config.getRootDir();

  const { type: typeFilter, slug: slugFilter, ...contentFilters } = filters;

  // Level 1, identify the files to fetch
  const filesToFetch: { type: string; slug: string }[] = [];

  const allTypes = await listDirCached(rootDir);
  const typeFilterPredicate = expressionToProvider(typeFilter);
  const typesToFetch = typeFilter
    ? allTypes.filter(typeFilterPredicate)
    : allTypes;

  await Promise.all(
    typesToFetch.map(async type => {
      const allSlugFileNames = await listDirCached(path.join(rootDir, type));
      const allSlugs = allSlugFileNames
        .filter(slug => slug.endsWith(".md"))
        .map(slug => slug.substring(0, slug.length - 3));
      const slugFilterPredicate = expressionToProvider(slugFilter);
      const slugsToFetch = slugFilter
        ? allSlugs.filter(slugFilterPredicate)
        : allSlugs;

      for (const slug of slugsToFetch) {
        filesToFetch.push({ type, slug });
      }
    })
  );

  // level 2, fetch & filter the contents from the selected files
  const allContents = await Promise.all(
    filesToFetch.map(async ({ type, slug }) => {
      return fetchContentCached(rootDir, type, slug);
    })
  );

  const contentPredicate = getContentFilterPredicate(contentFilters);

  const filteredContents = contentPredicate
    ? allContents.filter(contentPredicate)
    : allContents;

  // level 3, filter attributes
  const partialContents = filteredContents.map(content =>
    attributeFilter(content, attributes)
  );

  return partialContents;
};

const getContentFilterPredicate = (filters?: Record<string, AnyExpression>) => {
  const filterKeys = Object.keys(filters);

  if (filterKeys.length > 0) {
    const providers: Record<string, (data: unknown) => boolean> = {};
    filterKeys.forEach(key => {
      providers[key] = expressionToProvider(filters[key]);
    });

    return (content: Content) => {
      for (const key of filterKeys) {
        const provider = providers[key];
        if (!(provider && provider(content[key]))) {
          return false;
        }
      }
      return true;
    };
  }
};

const expressionToProvider = (expression: AnyExpression) => {
  if (expression === undefined) {
    return;
  }
  const fullExpression =
    expression === null ||
    typeof expression == "string" ||
    typeof expression == "number" ||
    typeof expression == "boolean"
      ? { eq: expression }
      : expression;

  const exprName = Object.keys(fullExpression)[0];
  const exprArg = Object.values(fullExpression)[0];
  const provider = Config.getFilterProvider(exprName);
  return (data: unknown) => {
    return provider(data, exprArg);
  };
};
