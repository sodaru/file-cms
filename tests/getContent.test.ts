import path from "path";
import { Config } from "../src/Config";
import { getContent } from "../src/getContent";

const commonRootDir = path.join(__dirname, "sample-data");

describe("getContent Tests", () => {
  test("Check title is returned", async () => {
    Config.setRootDir(commonRootDir);
    const blog1 = await getContent("blog", "blog-1");
    expect(blog1.title).toEqual("Understanding JavaScript Closures");
  });
  test("Check author is returned", async () => {
    Config.setRootDir(commonRootDir);
    const blog1 = await getContent("blog", "blog-1");
    expect(blog1.author).toEqual("Alice Smith");
  });
  test("Check date is returned", async () => {
    Config.setRootDir(commonRootDir);
    const blog1 = await getContent("blog", "blog-1");
    expect(blog1.date).toEqual("2024-08-12");
  });
  test("Check tags is returned", async () => {
    Config.setRootDir(commonRootDir);
    const blog1 = await getContent("blog", "blog-1");
    expect(blog1.tags).toEqual(["JavaScript", "Programming"]);
  });
  test("Check missing attributes", async () => {
    Config.setRootDir(commonRootDir);
    const content = await getContent("blog", "blog-1");
    expect(content).not.toHaveProperty("nonexistentAttribute");
  });
  test("Check extra attributes", async () => {
    Config.setRootDir(commonRootDir);
    const content = await getContent("blog", "blog-1");
    expect(content).toHaveProperty("type");
    expect(content).toHaveProperty("slug");
  });
  test("Check content for a different type", async () => {
    Config.setRootDir(commonRootDir);
    const content = await getContent("product", "product-1");
    expect(content).toHaveProperty("type", "product");
  });
  test("Check content for a different slug", async () => {
    Config.setRootDir(commonRootDir);
    const content = await getContent("service", "service-1");
    expect(content).toHaveProperty("slug", "service-1");
  });
  test("Handle non-existent type", async () => {
    Config.setRootDir(commonRootDir);
    await expect(getContent("files", "blog-1")).rejects.toThrow(
      "Content Not Found, rootDir = " +
        commonRootDir +
        ", type = files, slug = blog-1"
    );
  });
  test("Handle non-existent slug", async () => {
    Config.setRootDir(commonRootDir);
    await expect(getContent("blog", "blog-8")).rejects.toThrow("");
  });
  test("Check filtering specific attributes", async () => {
    Config.setRootDir(commonRootDir);
    const content = await getContent("blog", "blog-1", ["title", "author"]);
    expect(content).toEqual({
      title: "Understanding JavaScript Closures",
      author: "Alice Smith"
    });
  });
  test("Check filtering with no attributes", async () => {
    Config.setRootDir(commonRootDir);
    const content = await getContent("blog", "blog-1");
    expect(content).toEqual({
      type: "blog",
      slug: "blog-1",
      content:
        "# JavaScript Closures\n\nClosures are an important concept in JavaScript, enabling powerful patterns for scoping and memory management.",
      title: "Understanding JavaScript Closures",
      author: "Alice Smith",
      date: "2024-08-12",
      tags: ["JavaScript", "Programming"]
    });
  });
  test("Retrieve filtered content metadata", async () => {
    Config.setRootDir(commonRootDir);
    const content = await getContent("blog", "blog-1", ["author", "date"]);
    expect(content).toEqual({
      author: "Alice Smith",
      date: "2024-08-12"
    });
  });
  test("Check case sensitivity in attribute names", async () => {
    Config.setRootDir(commonRootDir);

    const blog1 = await getContent("blog", "blog-1", ["Title"]);
    expect(blog1.Title).toBeUndefined();
    const blog1Correct = await getContent("blog", "blog-1", ["title"]);
    expect(blog1Correct.title).toEqual("Understanding JavaScript Closures");
  });

  test("Check correct error for invalid file path", async () => {
    Config.setRootDir(commonRootDir);
    await expect(getContent("blog", "../tests/sample-data")).rejects.toThrow();
  });
  test("Check correct parsing when front matter is missing", async () => {
    Config.setRootDir(commonRootDir);
    const noFrontMatterContent = await getContent("blog", "blog-4");
    expect(noFrontMatterContent).toEqual({
      type: "blog",
      slug: "blog-4",
      content:
        "# Python for Data Science\n\nPython is the go-to programming language for data science due to its simplicity and the vast array of libraries available."
    });
  });
  test("Check content with special characters", async () => {
    Config.setRootDir(commonRootDir);
    const specialCharContent = await getContent("blog", "blog-5");
    expect(specialCharContent).toEqual({
      type: "blog",
      slug: "blog-5",
      title: "Special @#$% Characters!",
      author: "Sarah Brown",
      date: "2024-09-05",
      tags: ["Productivity", "Developers"],
      content:
        "# Productivity Tips\n\nThis content includes special characters: @#$%^&()."
    });
  });
  test("Check if title starts with a specific word", async () => {
    Config.setRootDir(commonRootDir);
    const content = await getContent("blog", "blog-1");
    expect((content.title as string).startsWith("Understanding")).toBe(true);
  });

  test("Check only metadata is returned when there is no content", async () => {
    Config.setRootDir(commonRootDir);
    const noContentMetadata = await getContent("service", "service-2");
    expect(noContentMetadata).toMatchSnapshot();
  });
  test("return full content when no attributes specified", async () => {
    Config.setRootDir(commonRootDir);
    const fullContent = await getContent("product", "product-1");
    expect(fullContent).toMatchSnapshot();
  });
});

describe("Get content With Seo", () => {
  test("with seo = true", async () => {
    Config.setRootDir(commonRootDir);
    await expect(
      getContent("blog", "blog-1", undefined, true)
    ).resolves.toMatchSnapshot();
  });

  test("with default value for non existing seo key", async () => {
    Config.setRootDir(commonRootDir);
    await expect(
      getContent("blog", "blog-1", undefined, {
        twitterSite: { default: "@xyz" }
      })
    ).resolves.toMatchSnapshot();
  });
});
