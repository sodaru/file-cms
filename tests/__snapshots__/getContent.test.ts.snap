// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check only metadata is returned when there is no content 1`] = `
{
  "category": "Marketing",
  "content": "",
  "priceRange": "$800 - $4000",
  "slug": "service-2",
  "tags": [
    "digital marketing",
    "services",
  ],
  "title": "Digital Marketing Strategy",
  "type": "service",
}
`;

exports[`return full content when no attributes specified 1`] = `
{
  "category": "Electronics",
  "content": "# SuperFast Laptop

A high-performance laptop designed for professionals who need speed and reliability.",
  "price": "$999",
  "releaseDate": "2024-07-10",
  "slug": "product-1",
  "tags": [
    "laptop",
    "electronics",
    "featured",
  ],
  "title": "SuperFast Laptop",
  "type": "product",
}
`;
